### PR TITLE
Test 1d

### PR DESCRIPTION
**Pull request by MuawizChaudhary**
This PR adds a whole suite of tests for the Torch, Tensorflow, and Numpy backend primitives. Additionally, tests for scattering and tests for backend primitives are broken into two separate files. Perhaps we want to do something similar with frontend tests? 

Additionally, some already existing backend functions have been modified to perform type checking, such as adding a complex check to the subsampling functions in the numpy and tensorflow backends.

### CODE REVIEW

**Comment by MuawizChaudhary** [comment]
we should really finish up the R2C pr first tho.

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 1]
```diff
@@ -28,9 +28,11 @@ def subsample_fourier(x, k):
         The input tensor periodized along the next to last axis to yield a
         tensor of size x.shape[-2] // k along that dimension.
     """
+    complex_check(x)
```
Comment (eickenberg) [inline kymatio/scattering1d/backend/numpy_backend.py:31]: Are these changes intended for this PR or is this a stray commit from the R2C PR?

Comment (MuawizChaudhary) [reply kymatio/scattering1d/backend/numpy_backend.py:31]: Intended for this PR.

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 2]
```diff
@@ -28,9 +28,11 @@ def subsample_fourier(x, k):
...
+
+    y = x.reshape(-1, k, x.shape[-1] // k)
 
-    N = x.shape[-1]
-    res = x.reshape(x.shape[:-1] + (k, N // k)).mean(axis=(-2,))
+    res = y.mean(axis=(-2,))
```
Comment (janden) [inline kymatio/scattering1d/backend/numpy_backend.py:35]: Why this change?

File: tests/scattering1d/test_numpy_backend_1d.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
+import pytest
+import numpy as np
+
+from kymatio.scattering1d.backend.numpy_backend import backend
+
+def test_subsample_fourier(random_state=42):
+    rng = np.random.RandomState(random_state)
+    J = 10
+    # 1d signal 
+    x = rng.randn(2, 2**J) + 1j * rng.randn(2, 2**J)
```
Comment (janden) [inline tests/scattering1d/test_numpy_backend_1d.py:10]: Spaces around `**` (also below).

File: tests/scattering1d/test_numpy_backend_1d.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            # of original array
+            for t in range(1, pad_right + 1):
+                assert np.allclose(x_pad[..., x_pad.shape[-1] - 1 - pad_right + t], x[..., x.shape[-1] - 1 - t])
+            # compare right part of padded array with right side of 
+            # original array
+            for t in range(1, pad_right + 1):
+                assert np.allclose(x_pad[..., x_pad.shape[-1] - 1 - pad_right - t], x[..., x.shape[-1] - 1 - t])
+
+    with pytest.raises(ValueError):
+        backend.pad(x, x.shape[-1], 0)
```
Comment (janden) [inline tests/scattering1d/test_numpy_backend_1d.py:49]: Check that the error messages are correct?

File: tests/scattering1d/test_numpy_backend_1d.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
+import pytest
+import numpy as np
+
+from kymatio.scattering1d.backend.numpy_backend import backend
+
```
Comment (janden) [inline tests/scattering1d/test_numpy_backend_1d.py:5]: Two blank lines between top-level definitions (PEP 8).

File: tests/scattering1d/test_numpy_backend_1d.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    assert np.allclose(x, z_1)
+
+    z_2 = backend.irfft(z)
+    assert not np.iscomplexobj(z_2)
+    assert np.allclose(x, z_2)
+
+    def coefficent(n):
+            return np.exp(-2 * np.pi * 1j * n)
+
+    x_r = np.random.rand(4)
```
Comment (janden) [inline tests/scattering1d/test_numpy_backend_1d.py:109]: Why are we doing this twice? Should be enough to just do it below.

File: tests/scattering1d/test_tensorflow_backend_1d.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    for j in range(J + 1):
+        x_f_sub = backend.subsample_fourier(x_f, 2**j)
+        x_sub = np.fft.ifft(x_f_sub, axis=-1)
+        assert np.allclose(x[:, ::2**j], x_sub)
+
+    with pytest.raises(TypeError) as te:
+        x_bad = x.real
+        backend.subsample_fourier(x_bad, 1)
+    assert "should be complex" in te.value.args[0]
+
```
Comment (janden) [inline tests/scattering1d/test_tensorflow_backend_1d.py:23]: Two blank lines between top-level definitions (PEP 8).

File: tests/scattering1d/test_numpy_backend_1d.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
+import pytest
+import numpy as np
+
+from kymatio.scattering1d.backend.numpy_backend import backend
+
+def test_subsample_fourier(random_state=42):
```
Comment (janden) [inline tests/scattering1d/test_numpy_backend_1d.py:6]: Remove this `random_state` business.

File: tests/scattering1d/test_tensorflow_backend_1d.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
+import pytest
+import numpy as np
+
+from kymatio.scattering1d.backend.tensorflow_backend import backend
+
+
+def test_subsample_fourier(random_state=42):
```
Comment (janden) [inline tests/scattering1d/test_tensorflow_backend_1d.py:7]: Remove `random_state`.

File: tests/scattering1d/test_tensorflow_backend_1d.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+            # compare right reflected part of padded array with right side
+            # of original array
+            for t in range(1, pad_right + 1):
+                assert np.allclose(x_pad[..., x_pad.shape[-1] - 1 - pad_right + t], x[..., x.shape[-1] - 1 - t])
+            # compare right part of padded array with right side of 
+            # original array
+            for t in range(1, pad_right + 1):
+                assert np.allclose(x_pad[..., x_pad.shape[-1] - 1 - pad_right - t], x[..., x.shape[-1] - 1 - t])
+
+    with pytest.raises(ValueError):
```
Comment (janden) [inline tests/scattering1d/test_tensorflow_backend_1d.py:49]: Check error messages.

File: tests/scattering1d/test_tensorflow_backend_1d.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    assert np.allclose(y, z)
+
+    z_1 = backend.ifft(z)
+    assert np.allclose(x, z_1)
+
+    z_2 = backend.irfft(z)
+    assert not np.iscomplexobj(z_2)
+    assert np.allclose(x, z_2)
+
+    def coefficent(n):
```
Comment (janden) [inline tests/scattering1d/test_tensorflow_backend_1d.py:107]: Again, once is enough.

File: kymatio/scattering1d/backend/numpy_backend.py (hunk 2 above)
Comment (MuawizChaudhary) [reply kymatio/scattering1d/backend/numpy_backend.py:35]: I wanted this to be more in line with the 2d and torch backends subsample fourier.

File: tests/scattering1d/test_numpy_backend_1d.py (hunk 4 above)
Comment (MuawizChaudhary) [reply tests/scattering1d/test_numpy_backend_1d.py:49]: Yeah you might have a point

File: tests/scattering1d/test_numpy_backend_1d.py (hunk 6 above)
Comment (MuawizChaudhary) [reply tests/scattering1d/test_numpy_backend_1d.py:109]: This is something I was wondering about. Thank you!

**Comment by MuawizChaudhary** [comment]
are we ready for merge?

File: kymatio/scattering1d/backend/numpy_backend.py (hunk 2 above)
Comment (janden) [reply kymatio/scattering1d/backend/numpy_backend.py:35]: Ok.

File: tests/scattering1d/test_numpy_backend_1d.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+        y = backend.ifft(x)
+    assert 'should be complex' in record.value.args[0]
+
+    with pytest.raises(TypeError) as record:
+        y = backend.irfft(x)
+    assert 'should be complex' in record.value.args[0]
+
+
+def test_fft():
+
```
Comment (janden) [inline tests/scattering1d/test_numpy_backend_1d.py:97]: Remove blank line.

File: tests/scattering1d/test_tensorflow_backend_1d.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    N = 128
+    x = np.random.rand(2, 4, N)
+
+    # similar to for loop in pad test
+    for pad_left in range(0, N - 16, 16):
+        pad_right = pad_left + 16
+        x_pad = backend.pad(x, pad_left, pad_right)
+        x_unpadded = backend.unpad(x_pad, pad_left, x_pad.shape[-1] - pad_right)
+        assert np.allclose(x, x_unpadded)
+
```
Comment (janden) [inline tests/scattering1d/test_tensorflow_backend_1d.py:76]: Two blank lines between top-level definitions.

Comment (MuawizChaudhary) [reply tests/scattering1d/test_tensorflow_backend_1d.py:76]: this is so weird theres two blank lines here in my vim editor.

File: tests/scattering1d/test_torch_backend_1d.py [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+    with pytest.raises(TypeError) as record:
+        y = backend.ifft(x)
+    assert 'should be complex' in record.value.args[0]
+
+    with pytest.raises(TypeError) as record:
+        y = backend.irfft(x)
+    assert 'should be complex' in record.value.args[0]
+
+def test_fft():
+    x = torch.randn(2, 1)
```
Comment (MuawizChaudhary) [inline tests/scattering1d/test_torch_backend_1d.py:208]: Remove.

File: tests/scattering1d/test_tensorflow_backend_1d.py (hunk 13 above)
Comment (janden) [reply tests/scattering1d/test_tensorflow_backend_1d.py:76]: There's only one blank line in my editor and on here.

**Comment by MuawizChaudhary** [comment]
Added!

**Comment by eickenberg** [comment]
yup seems good to go to me

**Comment by janden** [comment]
Let's make sure the tests pass, then we're ready to merge.

**Comment by MuawizChaudhary** [comment]
The tests pass on my end. I can try changing the tolerance and see if that changes things

**Comment by MuawizChaudhary** [comment]
Looks like jenkins passes.

**Comment by janden** [comment]
Still failing on Travis. Can you increase the tolerance for that test? Also make sure to put a comment explaining why we need the increased tolerances.
